### ISSUE TITLE
Momentary loss of RF link when changing flight mode

### ISSUE THREAD

**Issue by AJHAJHAJH** [EdgeTX/edgetx#5011]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Transmitter firmware

### Current Behavior

When selecting different flight modes, the RF link appears to drop maybe 1 time in 10.  the green LED on the Rx goes out, the red LED briefly illuminates, the servos go to failsafe positions and after a fraction of a second the RF link resumes, the red LED goes out and the green LED comes on.

This behaviour only affects my RMTX16s Mk1 and brand new Mk2 when using the internal MPM.

### Expected Behavior

The RF link should be constant through flight mode changes.

### Steps To Reproduce

1) Use a RadioMaster TX16s with an internal MPM to control a model with multiple flight modes.
2) Use  switch E to change flight modes repeatedly
3) Observe intermittent fail safes

### Version

2.10.0

### Transmitter

RadioMaster TX16S / TX16SMK2

### Anything else?

I have tried several different models using different FrSky receivers.  I have also tried a TX16sMk1 and a TX16sMk2.  The bug also shows if no servos are connected the the Rx.
All MPM's are V1.3.4.0 AETR FrSkyX2Cloned  and LBT. 

The bug  goes away if:
1) Use the same model setups on a Horus X10
2) Use an external  IRangex  MPM rather than the internal MPM of the TX16s
3) Flash the TX16s with OTX 2.3.14
[RM_MPM.zip](https://github.com/EdgeTX/edgetx/files/15317372/RM_MPM.zip)

**Comment by ParkerEde** [comment]
I cannot reproduce the problem described with my TX16S MK II and 2.10.0. I also switch the flightmodes with SE and as RX I use frsky ArcherPlus R6, SR6, X8R etc
Do you have any LUAs running?
Can you try a fresh default etx sdcard content and create a new default model and rebind the RX and try again?

**Comment by AJHAJHAJH** [comment]
"Do you have any LUAs running?"
Yes, and also widgets - disabling them makes no difference.


"Can you try a fresh default etx sdcard content and create a new default model and rebind the RX and try again?"

1) This worked correctly.  
2) I then copied one of my troublesome models (Stigra) to the fresh SD card. This also worked correctly.

3) I then copied the associated sound files to SOUNDS\en\Stigra - the problem reappeared
[StigraSounds.zip](https://github.com/EdgeTX/edgetx/files/15327267/StigraSounds.zip)

4) I selected external MPM rather than internal - the problem disappeared.

As an aside, prior to V2.8 I would always populate the model's sounds directory with .wav files for all modes, switches and logical switches.  V2.8 would take about a minute to select a model.  I'm not clever enough to understand the code in detail but it seemed that 2.7 looked for the file as needed whereas 2.8 checked each file found and generated a catalogue on model selection - which was what took all the time.  I have no idea if this is relevant.

**Comment by AJHAJHAJH** [comment]
After a sleepless night I have discovered:
1) This page https://github.com/EdgeTX/edgetx-sdcard-sounds says:
 "Note: EdgeTX supports up to 32khz"

2) If I resample my sounds down to 16kHz, the problem goes away.

I don't understand why this problem exists for the internal MPM but not the external one, but I have a solution.

Thank you for your time and effort, my apologies for wasting both.

**Comment by ParkerEde** [comment]
I think that the development should actually check why the RF link is temporarily lost only when working with unsuitable sound files. Something like this should be intercepted in any case, as it represents a potential risk. I would not say that this ticket should be closed, but that a developer should take a closer look at it.

**Comment by pfeerick** [comment]
Indeed... I think now that we have a possible lead as to the a specific reproducible cause, now need to see how to prevent/mitigate it. 

@AJHAJHAJH Thanks for narrowing that down, and uploading the files in https://github.com/EdgeTX/edgetx/issues/5011#issuecomment-2113521875 ... that should help narrowing it down.

**Comment by AJHAJHAJH** [comment]
"Out of curiosity, how where those generated ?"

The ones I uploaded were generated by an Android version of a voice from https://www.cereproc.com.  I wrote (badly) an App to generate files in whatever format Android wants and then convert them using  https://www.nch.com.au/switch/index.html

![NCHSwitch](https://github.com/EdgeTX/edgetx/assets/166018080/b7c4a1d6-0a8f-4818-96e6-387fc9708ce7)

**Comment by AJHAJHAJH** [comment]
I've messed with this some more:

I used FFMPEG to convert the files so that they are PcmWaveformat rather than WaveFormatEx  - no change.

I discovered that other files from  SD\SOUNDS\en\<ModelName> (L60-on, SB down etc.)  also cause the RF link to glitch, the problem is not confined to flight mode changes.

I renamed (three characters) the SD\SOUNDS\en\<ModelName>\<FlightMode> files and moved them to SD\SOUNDS\en\ and played them using a special function triggered by flight mode - no change.  

I then renamed  and moved **all** the files from SD\SOUNDS\en\<ModelName> to SD\SOUNDS\en\ and played them **all** using an appropriate special function and the problem went away.

Again, this problem only occurs for me when using an internal MPM in a TX16S.

I hope this information will help someone cleverer than me to locate and fix the problem.

**Comment by AJHAJHAJH** [comment]
Still experiencing this issue on 2.11.3, but have just discovered it does not affect my models which use only two servos (elevon equipped gliders).

**Comment by AJHAJHAJH** [comment]
Release 2.11.5 seems have made this issue better and worse at the same time.

It is better in that the loss of RF happens much less frequently.  Sometimes it can take many minutes to induce the fault.

It is worse in that the loss of RF is no longer momentary and now lasts for several seconds.  Compare the attached video with the one in the original post.

https://github.com/user-attachments/assets/a198e97f-e21d-4c6c-a7e9-06d6bd8b590d

**Comment by raphaelcoeffic** [comment]
@AJHAJHAJH would you mind testing with the builds from #7490? (don't forget to backup everything!) Let me know if that improves things.

**Comment by AJHAJHAJH** [comment]
I will gladly do some testing.  Assume I'm an idiot and help me find "the builds from #7490".  Presumably this refers to a firmware file somewhere.

Thanks,
AJH

**Comment by gismo2004** [comment]
@AJHAJHAJH https://github.com/EdgeTX/edgetx/actions/runs/28083017196/artifacts/7843301155 from here: https://github.com/EdgeTX/edgetx/actions/runs/28083017196?pr=7490

**Comment by AJHAJHAJH** [comment]
Back in May 2024, this bug made me think my TS16SMk1 was faulty, so I bought a new Mk2 which also displayed the bug.  This means the Mk1 became a spare and all my models were configured to use an external MPM on the Mk2.  Today, I did the following:

1) Update Mk1 to 12.2.2

2) Repeat the procedure which demonstrated the bug  except it didn't!  I tried with a GRX8 and XR8.  Both are now running Mike Blandford's UNI firmware which they weren't at the time I reported the bug.  Flashing back to FrSky firmware did not restore the bug.  Around two hours of running and flicking switches all worked flawlessly.

3) Flashed the 7490 firmware.  An hour of running and flicking switches, also flawless.

4) reformatted the SD card and put my archived 2.10.2 files on it and flashed 2.10.2 firmware.  The bug appeared in less than 15 seconds.


Summary: Both 2.12.2 and 7490 seem to work on the ground.

Next step?  Reconfigure a small, light model to use the internal MPM and test fly?

**Comment by AJHAJHAJH** [comment]
Continuing to test and came across an oddity with the 7490 firmware.

1) Model "Pupil" is selected
2) Move throttle to mid-point
3) Select Model "ServoTest" - I used this model for much of the testing of 7490 firmware
4) "Throttle not Idle, Press any key to skip" appears
5) Pressing any key has no effect(no beeps even)
6) Returning the throttle to idle has no effect
7) Sometimes, a long press (20 seconds) on the power button, followed by short presses shuts down the radio, sometimes it requires disconnecting the battery.

This is repeatable at will.

Tx battery is now flat, so adjourned until tomorrow.

AJH

[ServoTest.zip](https://github.com/user-attachments/files/29304451/ServoTest.zip)

[document1.zip](https://github.com/user-attachments/files/29304486/document1.zip)

**Comment by raphaelcoeffic** [comment]
@AJHAJHAJH This may or may not be related to the changes I did. Ideally, we'd compare what's comparable. So either you compare latest nightly with that PR, or I make you a patched version on top of 2.12.2.

Regarding what you just described with "throttle not idle", I believe this is most probably happening with nightly builds as well. Could you please retest with that?

**Comment by AJHAJHAJH** [comment]
Just flashed the nightly tx16s-0a65908.bin.  "Throttle not idle" issue is unchanged.  I will continue with the nightly to test for the RF link loss problem.

**Comment by raphaelcoeffic** [comment]
> Just flashed the nightly tx16s-0a65908.bin. "Throttle not idle" issue is unchanged. I will continue with the nightly to test for the RF link loss problem.

Created #7499 to tackle that particular issue (and https://github.com/EdgeTX/edgetx/pull/7498 to resolve it).

**Comment by AJHAJHAJH** [comment]
My Mk1 TS16S has just spent the last eight hours non stop running the 7490 firmware playing tracks from a model subdirectory without losing RF link and going into failsafe.  The original problem is either solved or hiding very well indeed.
